From 75a732b9bc3f9ffc9edb13464d63fb59d5b395b3 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Mon, 24 Jul 2023 20:39:43 +0800
Subject: [PATCH 26/53] [Huawei] mm/sharepool: Unify the process for unshare
 and sp_free

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Like sp_alloc and k2u, the processes of unshare and sp_free are similar.
First, get the spg, then get the spa, and lastly unmap the spa for each
task in the spg.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 114 +++++++++---------------------------------------
 1 file changed, 20 insertions(+), 94 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 45270b57a027..3cac9cea07a5 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1911,11 +1911,12 @@ static void sp_fallocate(struct sp_area *spa)
 		WARN(1, "sp fallocate failed %d\n", ret);
 }
 
-static int sp_free_inner(unsigned long addr, int spg_id)
+static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 {
 	int ret = 0;
 	struct sp_area *spa;
 	struct sp_group *spg;
+	const char *str = is_sp_free ? "sp_free" : "unshare_uva";
 
 	if (!current->mm) {
 		/*
@@ -1926,26 +1927,29 @@ static int sp_free_inner(unsigned long addr, int spg_id)
 
 	spg = sp_group_get_from_mm(current->mm, spg_id);
 	if (!spg) {
-		pr_err("sp free get group failed %d\n", spg_id);
+		pr_err("%s, get group failed %d\n", str, spg_id);
 		return -EINVAL;
 	}
 
 	spa = get_sp_area(spg, addr);
 	if (!spa) {
-		pr_debug("sp free invalid input addr %lx\n", addr);
+		pr_debug("%s, invalid input addr %lx\n", str, addr);
 		ret = -EINVAL;
 		goto out;
 	}
 
-	if (spa->type != SPA_TYPE_ALLOC) {
+	if ((is_sp_free && spa->type != SPA_TYPE_ALLOC) ||
+	    (!is_sp_free && spa->type == SPA_TYPE_ALLOC)) {
 		ret = -EINVAL;
-		pr_debug("sp free failed, %lx is not sp alloc addr\n", addr);
+		pr_warn("%s failed, spa_type is not correct\n", str);
 		goto drop_spa;
 	}
 
 	down_write(&spa->spg->rw_lock);
 	if (!spg_valid(spa->spg)) {
 		up_write(&spa->spg->rw_lock);
+		if (spa->type != SPA_TYPE_ALLOC)
+			vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL);
 		goto drop_spa;
 		/* we must return success(0) in this situation */
 	}
@@ -1953,7 +1957,7 @@ static int sp_free_inner(unsigned long addr, int spg_id)
 	/* the life cycle of spa has a direct relation with sp group */
 	if (unlikely(spa->is_dead)) {
 		up_write(&spa->spg->rw_lock);
-		pr_err_ratelimited("unexpected double sp free\n");
+		pr_err_ratelimited("%s, unexpected double sp free\n", str);
 		dump_stack();
 		ret = -EINVAL;
 		goto drop_spa;
@@ -1963,7 +1967,12 @@ static int sp_free_inner(unsigned long addr, int spg_id)
 
 	down_read(&spa->spg->rw_lock);
 	__sp_free(spa->spg, spa->va_start, spa_size(spa), NULL);
-	sp_fallocate(spa);
+	if (spa->type == SPA_TYPE_ALLOC) {
+		sp_fallocate(spa);
+	} else {
+		vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL);
+		spa->kva = 0;
+	}
 	up_read(&spa->spg->rw_lock);
 
 	if (current->mm != NULL)
@@ -1998,7 +2007,7 @@ int mg_sp_free(unsigned long addr, int id)
 	if (current->flags & PF_KTHREAD)
 		return -EINVAL;
 
-	return sp_free_inner(addr, id);
+	return sp_free_inner(addr, id, true);
 }
 EXPORT_SYMBOL_GPL(mg_sp_free);
 
@@ -2886,90 +2895,6 @@ void *mg_sp_make_share_u2k(unsigned long uva, unsigned long size, int tgid)
 }
 EXPORT_SYMBOL_GPL(mg_sp_make_share_u2k);
 
-/*
- * sp_unshare_uva - unshare a uva from sp_make_share_k2u
- * @uva: the uva to be unshared
- * @size: not used actually
- * @spg_id: specify the spg of the uva; for local group, it can be SPG_ID_DEFAULT
- *          unless current process is exiting.
- *
- * Procedure of unshare uva must be compatible with:
- *
- * 1. DVPP channel destroy procedure:
- * do_exit() -> exit_mm() (mm no longer in spg and current->mm == NULL) ->
- * exit_task_work() -> task_work_run() -> __fput() -> ... -> vdec_close() ->
- * sp_unshare(uva, local_spg_id)
- */
-static int sp_unshare_uva(unsigned long uva, unsigned long size, int spg_id)
-{
-	int ret = 0;
-	struct sp_area *spa;
-	struct sp_group *spg;
-
-	if (!current->mm) {
-		/*
-		 * XXXX: should process this branch carefully instead of return an error.
-		 */
-		return -EINVAL;
-	}
-
-	spg_id = (spg_id == SPG_ID_NONE) ? SPG_ID_DEFAULT : spg_id;
-	spg = sp_group_get_from_mm(current->mm, spg_id);
-	if (!spg) {
-		pr_err("sp free get group failed %d\n", spg_id);
-		return -EINVAL;
-	}
-
-	/* All the spa are aligned to 2M. */
-	spa = get_sp_area(spg, ALIGN_DOWN(uva, PMD_SIZE));
-	if (!spa) {
-		ret = -EINVAL;
-		pr_err("invalid input uva %lx in unshare uva\n", (unsigned long)uva);
-		goto out;
-	}
-
-	if (spa->type != SPA_TYPE_K2TASK && spa->type != SPA_TYPE_K2SPG) {
-		pr_err("unshare wrong type spa\n");
-		ret = -EINVAL;
-		goto out_drop_area;
-	}
-
-	down_write(&spa->spg->rw_lock);
-	if (!spg_valid(spa->spg)) {
-		up_write(&spa->spg->rw_lock);
-		pr_info("no need to unshare uva, sp group of spa is dead\n");
-		goto out_clr_flag;
-	}
-	/* the life cycle of spa has a direct relation with sp group */
-	if (unlikely(spa->is_dead)) {
-		up_write(&spa->spg->rw_lock);
-		pr_err("unexpected double sp unshare\n");
-		dump_stack();
-		ret = -EINVAL;
-		goto out_drop_area;
-	}
-	spa->is_dead = true;
-	up_write(&spa->spg->rw_lock);
-
-	down_read(&spa->spg->rw_lock);
-	__sp_free(spa->spg, spa->va_start, spa->real_size, NULL);
-	up_read(&spa->spg->rw_lock);
-
-	if (current->mm != NULL)
-		sp_update_process_stat(current, false, spa);
-
-out_clr_flag:
-	if (!vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL))
-		pr_info("clear spa->kva %ld is not valid\n", spa->kva);
-	spa->kva = 0;
-
-out_drop_area:
-	__sp_area_drop(spa);
-out:
-	sp_group_put(spg);
-	return ret;
-}
-
 /* No possible concurrent protection, take care when use */
 static int sp_unshare_kva(unsigned long kva, unsigned long size)
 {
@@ -3037,8 +2962,9 @@ int mg_sp_unshare(unsigned long va, unsigned long size, int spg_id)
 		return -EINVAL;
 
 	if (va < TASK_SIZE) {
-		/* user address */
-		ret = sp_unshare_uva(va, size, spg_id);
+		/* All the spa are aligned to 2M. */
+		spg_id = (spg_id == SPG_ID_NONE) ? SPG_ID_DEFAULT : spg_id;
+		ret = sp_free_inner(ALIGN_DOWN(va, PMD_SIZE), spg_id, false);
 	} else if (va >= PAGE_OFFSET) {
 		/* kernel address */
 		ret = sp_unshare_kva(va, size);
-- 
2.17.1

